### ISSUE TITLE
Kron with gradient for sparse tensors

### ISSUE THREAD

**Issue by SnoodMcGuinness** [pytorch/pytorch#134069]
### 🚀 The feature, motivation and pitch

I am using torch as part of some code that solves quantum systems in a differentiable way. An important part of this process is constructing large sparse matrices, often via the kronecker product of a number of matrices. In order for gradients to propagate from this large sparse matrix back to the parameters used in their construction, there needs to be a kron function that operates on sparse matrices and propagates gradients.

### Alternatives

It is unfeasible to first kron while dense and then convert to sparse as the full dense matrices do not fit in memory.
I can implement a kronecker function by calculating the new coordinates and values and creating a new sparse tensor with them, but in the creation of the new tensor the computation tree is broken:
```
def sparse_kron(input: torch.Tensor, other: torch.Tensor):
    # This version is a little weird but should propagate gradients
    assert input.ndim == other.ndim
    input_indices = input.indices()
    other_indices = other.indices()

    new_indices = []
    for input_idx in input_indices.T:
        for other_idx in other_indices.T:
            new_indices.append(
                [input_idx[i] * other.shape[i] + other_idx[i] for i in range(input.ndim)])


    new_indices = torch.tensor(new_indices, requires_grad=False).T

    new_values = torch.kron(input.values(), other.values())

    if new_indices.ndim == 1:
        new_indices = new_indices.reshape([input.ndim, 0])

    new_shape = [n * m for n, m in zip(input.shape, other.shape)]

    return torch.sparse_coo_tensor(new_indices, new_values, new_shape, dtype=input.dtype, device=input.device, is_coalesced=True)
```
fails to propagate the gradients.

**Comment by SnoodMcGuinness** [comment]
Apologies for my mistake, it seems the code above does indeed propagate the gradients I need:
```
t = torch.tensor([0.5], requires_grad=True)

cdag = torch.tensor([[0., 1.], [0., 0.]], requires_grad=False).to_sparse()
c = torch.tensor([[0., 0.], [1., 0.]], requires_grad=False).to_sparse()
cnum = torch.tensor([[0., 0], [0., 1.]], requires_grad=False).to_sparse()
eye = torch.eye(2, requires_grad=False).to_sparse()

H = t * sparse_kron(cdag, c) + t * sparse_kron(c, cdag)
loss = H.sum()
loss.backward()
print(t.grad)

t.grad = None

H = sparse_kron(t * cdag, c) + t * sparse_kron(t * c, cdag)
loss = H.sum()
loss.backward()
print(t.grad)
```
returns
`tensor([2.])
tensor([2.])`

and more generally, 
```
a_values = torch.rand([2,1,1], requires_grad=True) # values want to keep track of gradient
a = a_values.to_sparse()

# Introduce some large sparse matrices to kron with
b_values = torch.rand([10,10,10])
b_sparsity = torch.rand([10,10,10]) > 0.99
b = (b_values * b_sparsity).to_sparse()

c_values = torch.rand([10,10,10])
c_sparsity = torch.rand([10,10,10]) > 0.99
c = (c_values * c_sparsity).to_sparse()

d_values = torch.rand([10,10,10])
d_sparsity = torch.rand([10,10,10]) > 0.99
d = (d_values * d_sparsity).to_sparse()

# Kron to get a very large sparse array without ever getting dense representation
large_array = sparse_kron(a, sparse_kron(b, sparse_kron(c, d)))
print("Dense elements:", large_array.numel(), "Sparse elements:", large_array._nnz())

loss = torch.sum(large_array)
loss.backward()
print(a_values.grad)
```
returns
```
Dense elements: 2000000000 Sparse elements: 1296
tensor([[[69.6925]], [[69.6925]]])
```
So this does work as expected for now. Though I still think a native implementation would be useful in the future!

**Comment by SnoodMcGuinness** [comment]
On the offchance that anyone uses this in the future, the loops in the version I posted are pretty slow. I have experimented speeding the generation of the indices up with torch.repeat and torch.expand:

```
def old_index_combination(input, other):
    input_indices = input.indices()
    other_indices = other.indices()
    new_indices = []
    for input_idx in input_indices.T:
        for other_idx in other_indices.T:
            new_indices.append(
                [input_idx[i] * other.shape[i] + other_idx[i] for i in range(input.ndim)])


    new_indices = torch.tensor(new_indices, requires_grad=False).T
    return new_indices

def new_index_combination_repeat(input, other):
    input_indices = input.indices()
    other_indices = other.indices()
    
    repeated_input_indices = (torch.repeat_interleave(input_indices, other_indices.shape[1], dim=-1).T * torch.tensor(other.shape)).T
    repeated_other_indices = other_indices.repeat((1, input_indices.shape[1]))
    new_indices = repeated_input_indices + repeated_other_indices
    
    return new_indices

def new_index_combination_expand(input, other):
    input_indices = input.indices()
    other_indices = other.indices()
    
    input_indices_expanded = input_indices.expand(other_indices.shape[1], *input_indices.shape).T * torch.tensor(other.shape).reshape(1,-1,1)
    other_indices_expanded = other_indices.expand(input_indices.shape[1], *other_indices.shape)
    new_indices = torch.permute(input_indices_expanded + other_indices_expanded, (1,0,2)).reshape(2,-1)

    return new_indices
```

and tested the speed:
```
size = 1000
nelem = 240

values = torch.rand(nelem, requires_grad=False)
indices = torch.randint(0, size, size=(2, nelem))
shape = [size, size]
a = torch.sparse_coo_tensor(indices, values, shape).coalesce()

size = 5000
nelem = 325
values = torch.rand(nelem, requires_grad=False)
indices = torch.randint(0, size, size=(2, nelem))
shape = [size, size]
b = torch.sparse_coo_tensor(indices, values, shape).coalesce()
```
```
old_index_combinations = old_index_combination(a, b)
print(torch.all(old_index_combinations == new_index_combination_expand(a, b)))
print(torch.all(old_index_combinations == new_index_combination_repeat(a, b)))
```
--->
```
tensor(True)
tensor(True)
```
```
%time old_index_combination(a, b)
%timeit new_index_combination_repeat(a, b)
%timeit new_index_combination_expand(a, b)
```
--->
```
CPU times: user 4.38 s, sys: 254 ms, total: 4.63 s
Wall time: 4.03 s
202 μs ± 14.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
141 μs ± 4.31 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

with a new sparse kron function being:
```
def sparse_kron(input: torch.Tensor, other: torch.Tensor):
    assert input.ndim == other.ndim
    input_indices = input.indices()
    other_indices = other.indices()

    input_indices_expanded = input_indices.expand(other_indices.shape[1], *input_indices.shape).T * torch.tensor(other.shape).reshape(1,-1,1)
    other_indices_expanded = other_indices.expand(input_indices.shape[1], *other_indices.shape)
    new_indices = torch.permute(input_indices_expanded + other_indices_expanded, (1,0,2)).reshape(2,-1)

    new_values = torch.kron(input.values(), other.values())

    if new_indices.ndim == 1:
        new_indices = new_indices.reshape([input.ndim, 0])

    new_shape = [n * m for n, m in zip(input.shape, other.shape)]

    return torch.sparse_coo_tensor(new_indices, new_values, new_shape, dtype=input.dtype, device=input.device, is_coalesced=True)
```

**Comment by amjames** [comment]
Hi @SnoodMcGuinness the python implementation you posted seems to be a notable improvement over the original version, if that meets your needs it would probably be the best solution for you. 

Another route you could explore is contributing a kernel for the op which handles sparse inputs, if you go this route you could add me as a reviewer.
